include/merlin/core_kernels.cuh

```diff
@@ -1909,162 +1898,206 @@ void read_or_write_by_cpu(V** __restrict table_value_addrs,
   }
 }
 
-/* find or insert with the end-user specified meta.
+/*
+ * find or insert with IO operation. This kernel is
+ * usually used for the pure HBM mode for better performance.
  */
 template <class K, class V, class M, uint32_t TILE_SIZE = 4>
-__global__ void find_or_insert_kernel(
-    const Table<K, V, M>* __restrict table, const K* __restrict keys,
-    V** __restrict vectors, M* __restrict metas, bool* __restrict found,
-    int* __restrict keys_index, const size_t N) {
-  Bucket<K, V, M>* buckets = table->buckets;
-  int* buckets_size = table->buckets_size;
-  const size_t bucket_max_size = table->bucket_max_size;
-  const size_t buckets_num = table->buckets_num;
-  const size_t dim = table->dim;
-
-  size_t tid = (blockIdx.x * blockDim.x) + threadIdx.x;
+__global__ void find_or_insert_kernel_with_io(
+    const Table<K, V, M>* __restrict table, const size_t bucket_max_size,
+    const size_t buckets_num, const size_t dim, const K* __restrict keys,
+    V* __restrict values, M* __restrict metas, const size_t N) {
   auto g = cg::tiled_partition<TILE_SIZE>(cg::this_thread_block());
-  int rank = g.thread_rank();
-  Bucket<K, V, M>* bucket;
-  unsigned found_vote;
+  int* buckets_size = table->buckets_size;
 
-  for (size_t t = tid; t < N; t += blockDim.x * gridDim.x) {
+  for (size_t t = (blockIdx.x * blockDim.x) + threadIdx.x; t < N;
+       t += blockDim.x * gridDim.x) {
     int key_pos = -1;
-    size_t key_idx = t / TILE_SIZE;
-    int local_size = 0;
+    const size_t key_idx = t / TILE_SIZE;
```

Comment:
In general case, the KV number in a batch is less than 2^31. If not, there will be too many keys in our table.
So can we use the uint32_t instead of size_t? The same as the variable in the function of find_key_position .

Reply:
Make sense, will do some optimization in a seperate PR.